### ISSUE TITLE
Initial version for an API to configure block height of a transaction

### ISSUE THREAD

**Issue by digorithm** [FuelLabs/fuels-rs#410]
# Problem

Recently, some people expressed the need for a way to configure the target block height for a transaction, i.e., configuring _when_ a transaction should be executed. 

In Ethereum/Solidity, one can do something like this:

```js
contract.methodName({ blockNumber: blockNumber }).then((result) => {
    console.log(result);
});
```

In order to execute the call at `blockNumber`. 

In addition to setting the block height for a transaction, we would also like a way to quickly get the current block height, which is currently but you have to bypass the SDK's API and use `fuel-core`'s `Client` directly. This could be simplified by creating a shallow wrapper around the `Client`'s block API in the SDK's `Provider` struct.

So, three APIs would be nice to have:

- [ ] Get the current block height using the SDK's API
- [ ] `tx.set_block_height(block_height: u64)` to set the target block height for the transaction `tx`
- [ ] `add_to_block_height(blocks_to_add: u64)` force `blocks_to_add` blocks in front of the current block height

# Blockers

Given the current state of the client (`fuel-core`), it's not possible yet to _properly_ do the setting of block height in a transaction. For this to be possible, it needs the txpool to be fully released. This will enable transactions to sit in the txpool queue until they can be included at a specific block. But currently, it doesn't work because blocks are only produced by submitting new transactions.

# Design

Even though there's this big blocker, we can create the initial APIs for these functionalities and make it work by using a "brute-force" design; an inefficient way to get this to work while we wait for the client to deliver the txpool and some other features to support the proper way of doing this.

It would work as follows. `tx.set_block_height(block_height: u64)` would, internally, create $n$ dummy transactions where $n$ is $blockHeight - currentBlockHeight$. Because 1 transaction executed == one block produced, running these $n$ transactions before running the actual `tx` would make the `tx` be executed at the desired target `block_height`. For `get_block_height`, getting the current block height using the `Client`'s API is relatively trivial, so that would be a simple wrapper.

When the blockers from the client are gone, i.e., txpool released and support for respecting the `block_height` / `maturity` of a tx before including it in the pool, we can do follow-up work to re-work the internals of these new APIs (`set_block_height`, etc.) by doing it in a proper way, using the new `Client`'s API for that.

**Comment by adlerjohn** [comment]
Is that actually how ethers-rs does it? I was under the assumption you could use the `evm_mine` RPC endpoint (wrapped in an SDK function of course) to advance block height.

**Comment by digorithm** [comment]
> ```js
> contract.methodName({ blockNumber: blockNumber }).then((result) => {
>     console.log(result);
> });
> ```

Not sure, it's on my list to check how they do it internally. I was just checking what the `ethers.js` (_JS_) API for that look like (without checking the impl). 

But yeah, I don't _think_ that's how it's done internally because they have a more mature client in terms of these kinds of nice utilities to play with blocks (`evm_mine` as you said), etc.

That's why at the end I mention that we should re-work it so that we use these upcoming `fuel-core`'s API to do this kind of stuff -- I recently chatted with @Voxelot about it and it's on their roadmap, mainly blocked by the txpool.

This initial version would be just to give some sort of "relief" to some users that want to test their contract calls w.r.t different target block heights.

**Comment by adlerjohn** [comment]
Can you link to the relevant sections of ethers-rs where block height of a transaction is set?

**Comment by digorithm** [comment]
Yeah, I was just taking a look at how the js one does it because that was the reference API that was given to me when they asked if we can do something like that 😂. 

But a quick look on `ethers-rs` I found this:

The method that sets the block to a contract call (eventually it'll be a tx): https://github.com/gakonst/ethers-rs/blob/a532eb4d29c963f5b7baa404c3a821e26561ed94/ethers-contract/src/call.rs#L78

```Rust
/// Sets the `block` field for sending the tx to the chain
pub fn block<T: Into<BlockId>>(mut self, block: T) -> Self {
    self.block = Some(block.into());
    self
}
```

The call itself takes a block: https://github.com/gakonst/ethers-rs/blob/05864ff9ed6382d26825763c977eb03647a9d484/ethers-providers/src/lib.rs#L327-L333

```Rust
async fn call(
    &self,
    tx: &TypedTransaction,
    block: Option<BlockId>,
) -> Result<Bytes, Self::Error> {
    self.inner().call(tx, block).await.map_err(FromErr::from)
}
```

Right now, in the client, you can't pass an arbitrary block. I mean, you can set the maturity of the tx, which I guess it'll _eventually_ be used by the txpool to decide when to include the tx in a given block based on the maturity and the current block height?

**Comment by adlerjohn** [comment]
> which I guess it'll eventually be used by the txpool to decide when to include the tx in a given block based on the maturity and the current block height?

There's no guarantee the transaction will be included, and if it is included then it's only guaranteed to be _at least_ at the maturity height.

**Comment by digorithm** [comment]
> > which I guess it'll eventually be used by the txpool to decide when to include the tx in a given block based on the maturity and the current block height?
> 
> There's no guarantee the transaction will be included (blocks may be full of other transactions), and if it is included then it's only guaranteed to be _at least_ at the maturity height.

Ah, I understand. How would the user set the desired block height for a tx work in this setting then? In the current situation (1 tx == 1 block) it seems straightforward, but it isn't very clear (to me) how it would work when the txpool would be doing its thing.

**Comment by adlerjohn** [comment]
You would have to make sure the node is running in a way where the SDK can control the block contents, rather than relying on some default txpool logic.

**Comment by segfault-magnet** [comment]
IIUC guaranteeing that our transaction will be in the n-th block is only possible if we have some intimate control over the node -- i.e. this can only ever be done in a test environment?

Otherwise (in production if you will) best you can do is guarantee it will not be before the n-th block by setting the maturity to n (which we can already do).

Asking because I'm not sure where to place the new API. At a glance, I can't find anything in our contract call setup that implies a test environment -- hate to make this the first thing.

Looking at the [link](https://github.com/gakonst/ethers-rs/blob/05864ff9ed6382d26825763c977eb03647a9d484/ethers-providers/src/lib.rs#L327-L333) above it seems that ethers-rs puts this functionality right in the middle of the 'usable-in-a-non-test-environment' API.

Makes me either doubt this is test-only or doubt whether we want to mix test-only and production-usable APIs?

**Comment by adlerjohn** [comment]
> this can only ever be done in a test environment?

Correct, that's what I was alluding to above.

> So my thinking is we don't do the 'set_block_height' on the tx level but leave that to maturity. But rather do a test helper fn add_blocks which will add the blocks, after which, due to the matching maturity and no other concurrent tx, our tx should execute at height n.

I agree with this for the reasons you outlined: a clear separation between test-only functionality and everywhere-functionality.

**Comment by segfault-magnet** [comment]
Currently having issues with using `maturity`. More [here](https://github.com/FuelLabs/fuel-core/issues/432).

Will proceed on the helper fn for achieving block height until this is resolved.
